MediaBrowser.MediaEncoding/Attachments/AttachmentExtractor.cs

```diff
@@ -0,0 +1,278 @@
+using Microsoft.Extensions.Logging;
+using UtfUnknown;
+
+namespace MediaBrowser.MediaEncoding.Attachments
+{
+    public class AttachmentExtractor : IAttachmentExtractor
+    {
+        private readonly ILibraryManager _libraryManager;
+        private readonly ILogger _logger;
+        private readonly IApplicationPaths _appPaths;
+        private readonly IFileSystem _fileSystem;
+        private readonly IMediaEncoder _mediaEncoder;
+        private readonly IMediaSourceManager _mediaSourceManager;
+        private readonly IProcessFactory _processFactory;
+
+        public AttachmentExtractor(
+            ILibraryManager libraryManager,
+            ILoggerFactory loggerFactory,
+            IApplicationPaths appPaths,
+            IFileSystem fileSystem,
+            IMediaEncoder mediaEncoder,
+            IMediaSourceManager mediaSourceManager,
+            IProcessFactory processFactory)
+        {
+            _libraryManager = libraryManager;
+            _logger = loggerFactory.CreateLogger(nameof(AttachmentExtractor));
+            _appPaths = appPaths;
+            _fileSystem = fileSystem;
+            _mediaEncoder = mediaEncoder;
+            _mediaSourceManager = mediaSourceManager;
+            _processFactory = processFactory;
+        }
+
+        private string AttachmentCachePath => Path.Combine(_appPaths.DataPath, "attachments");
+
+        public async Task<(MediaAttachment attachment, Stream stream)> GetAttachment(BaseItem item, string mediaSourceId, int attachmentStreamIndex, CancellationToken cancellationToken)
+        {
+            if (item == null)
+            {
+                throw new ArgumentNullException(nameof(item));
+            }
+            if (string.IsNullOrWhiteSpace(mediaSourceId))
+            {
+                throw new ArgumentNullException(nameof(mediaSourceId));
+            }
+
+            var mediaSources = await _mediaSourceManager.GetPlayackMediaSources(item, null, true, false, cancellationToken).ConfigureAwait(false);
+            MediaSourceInfo mediaSource;
+            MediaAttachment mediaAttachment;
+            try
+            {
+                mediaSource = mediaSources
+                    .First(i => string.Equals(i.Id, mediaSourceId, StringComparison.OrdinalIgnoreCase));
+            }
+            catch (Exception ex) when (ex is ArgumentNullException || ex is InvalidOperationException)
+            {
+                throw new ResourceNotFoundException($"MediaSource {mediaSourceId} not found");
+            }
+            try
+            {
+                mediaAttachment = mediaSource.MediaAttachments
+                    .First(i => i.Index == attachmentStreamIndex);
+            }
+            catch (Exception ex) when (ex is ArgumentNullException || ex is InvalidOperationException)
+            {
+                throw new ResourceNotFoundException($"MediaSource {mediaSourceId} has no attachment with stream index {attachmentStreamIndex}");
+            }
```

Comment:
Please don't ever do this, use `FirstOrDefault` and check if the result is null.